collector/internal/controller.go

```diff
@@ -0,0 +1,243 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/ebpf-profiler/collector/internal"
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/consumerprofiles"
+	"go.uber.org/zap"
+
+	hostinfo "go.opentelemetry.io/ebpf-profiler/host"
+	"go.opentelemetry.io/ebpf-profiler/reporter"
+	"go.opentelemetry.io/ebpf-profiler/times"
+	"go.opentelemetry.io/ebpf-profiler/tracehandler"
+	"go.opentelemetry.io/ebpf-profiler/tracer"
+	tracertypes "go.opentelemetry.io/ebpf-profiler/tracer/types"
+	"go.opentelemetry.io/ebpf-profiler/util"
+)
+
+var (
+	agentShutdown = func() {}
+)
+
+type Controller struct {
+	host         component.Host
+	cancel       context.CancelFunc
+	logger       *zap.Logger
+	nextConsumer consumerprofiles.Profiles
+	config       *Config
+}
+
+func NewController(logger *zap.Logger, nextConsumer consumerprofiles.Profiles,
+	cfg *Config) *Controller {
+	return &Controller{
+		logger:       logger,
+		nextConsumer: nextConsumer,
+		config:       cfg,
+	}
+}
+
+// Start starts the receiver.
+func (c *Controller) Start(ctx context.Context, host component.Host) error {
+	cfg := c.config
```

Comment:
Since there's a lot of duplication with startup logic in main here, maybe we should factor out commonalities into a separate package (e.g. `agent`) that we can reuse here and in `main`.

Reply:
While your point definitely makes sense, this is far from being a trivial change in `main.go`, as it mixes a lot of things together. So making that refactoring looks like it's going to take several PRs.
While I'm happy to look into that, I wonder if we shouldn't start without this refactoring, so we can move forward with the collector support.